### PR TITLE
[DatePicker] reuse existing components in caption

### LINKED ISSUE #2801: [DateTime] Drop dependency on Icon fonts
#### Environment

- __Package version(s)__: @blueprintjs/datetime": "3.0.0"
- __Browser and OS versions__: Chromium & Ubuntu 16.04

#### Feature request
The current implementation of the Date picker relies on the icon font for the previous and next month buttons and as far as I understood, icon fonts were considered legacy within this project. It would be nice to be able have SVG icons for those, or even better, be able to provide our own (for example, FontAwesome icons).
Thanks for your consideration!

### PR DESCRIPTION
**Pull request by giladgray**
#### Changes proposed in this pull request:

- use minimal `HTMLSelect` for month/year dropdowns.
- use minimal `Button` for next/previous buttons.
- use `navbarElement` prop to render custom next/prev buttons instead of restyling the existing classes.
    - caption is responsible for month/year selects
    - navbar is responsible for prev/next buttons.
- `DatePicker` uses single `handleMonthChange` for all DayPicker events
    - caption (month/year select) and navbar (prev/next buttons) now all use the same logic for changes!
- use `Divider` for all the internal separators
- add `HTMLSelect` `iconProps` prop so we can position the icon

#### Screenshot

`DatePicker`
datepicker-caption

`DateRangePicker`
daterangepicker-caption

`DateRangePicker contiguousCalendarMonths`
daterangepicker-caption-no-cont

### CODE REVIEW

**Comment by giladgray** [comment]
#2793 will fix the failing `DRP` tests here.

File: packages/datetime/src/datePickerCaption.tsx [hunk 1]
```diff
@@ -44,73 +47,51 @@ export class DatePickerCaption extends React.PureComponent<IDatePickerCaptionPro
...
-                    {monthOptionElements}
-                </select>
-                <Icon
-                    className={Classes.DATEPICKER_CAPTION_CARET}
-                    icon="caret-down"
-                    style={{ left: this.state.monthWidth + 5 }}
-                />
-            </div>
+            <HTMLSelect
+                iconProps={{ style: { left: this.state.monthWidth } }}
```
Comment (cmslewis) [inline packages/datetime/src/datePickerCaption.tsx]: Interesting. What's going on here? An icon is absolutely positioned inside the `<select>`? Oh, or is this referring to the caret? Can we just let the select morph to the width of the widest `<option>`, or does our style customization make that tricky?

__EDIT:__ Took a look at the GIFs. Looks odd to me when the caret is right next to the text but the hoverable area goes way beyond the icon to the right. Can we just follow the style of default `<select>`s and leave the caret flush on the right side of the hoverable region?

File: packages/datetime/src/datePickerCaption.tsx [hunk 2]
```diff
@@ -123,25 +104,29 @@ export class DatePickerCaption extends React.PureComponent<IDatePickerCaptionPro
...
     private positionArrows() {
         // measure width of text as rendered inside our container element.
-        const monthWidth = Utils.measureTextWidth(
+        const monthWidth = measureTextWidth(
             this.displayedMonthText,
             Classes.DATEPICKER_CAPTION_MEASURE,
             this.containerElement,
         );
-        this.setState({ monthWidth });
+        if (monthWidth > this.containerElement.firstElementChild.clientWidth - Icon.SIZE_STANDARD) {
```
Comment (cmslewis) [inline packages/datetime/src/datePickerCaption.tsx]: ~Can you post some screenshots here to demonstrate each case by example?~ Never mind.

File: packages/datetime/src/datePickerNavbar.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2015 Palantir Technologies, Inc. All rights reserved.
```
Comment (cmslewis) [inline packages/datetime/src/datePickerNavbar.tsx]: 2018?

File: packages/datetime/src/datePickerNavbar.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import classNames from "classnames";
+import * as Classes from "./common/classes";
+import { areSameMonth } from "./common/dateUtils";
+
+export interface IDatePickerNavbarProps extends NavbarElementProps {
+    maxDate: Date;
+    minDate: Date;
+}
+
+export class DatePickerNavbar extends React.PureComponent<IDatePickerNavbarProps> {
```
Comment (cmslewis) [inline packages/datetime/src/datePickerNavbar.tsx]: Nice.

File: packages/datetime/src/datePickerNavbar.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    public render() {
+        const { classNames: classes, month, maxDate, minDate } = this.props;
+
+        return (
+            <div className={classNames(Classes.DATEPICKER_NAVBAR, classes.navBar)}>
```
Comment (cmslewis) [inline packages/datetime/src/datePickerNavbar.tsx]: Where does `classes.navBar` come from / what is its value? Not seeing it in the changed lines of the parent component(s). I assume it's just the `ReactDayPicker` default, and we need to provide it in order for basic ReactDayPicker styles to be applied in addition to our style overrides?

File: packages/datetime/test/dateInputTests.tsx [hunk 6]
```diff
@@ -557,6 +564,7 @@ describe("<DateInput>", () => {
 
             wrapper
                 .find(`.${Classes.DATEPICKER_MONTH_SELECT}`)
+                .find("select")
                 .simulate("change", { value: Months.FEBRUARY.toString() });
```
Comment (cmslewis) [inline packages/datetime/test/dateInputTests.tsx]: Maybe add a `changeSelectOption()` helper?

```tsx
function changeSelectOption(wrapper: ReactWrapper<any, any>, selectCssClass: string, value: string) {
    wrapper
        .find(`.${selectCssClass}`)
        .find("select")
        .simulate("change", { value });
}
```

Or more generically:

```tsx
function changeSelectOption(wrapper: ReactWrapper<any, any>, selectSelector: string, value: string) {...}
```

File: packages/datetime/src/datePickerCaption.tsx (hunk 1 above)
Comment (giladgray) [reply packages/datetime/src/datePickerCaption.tsx]: @cmslewis i've got back and forth on this myself. you can disable the `left` property in the browser. i think it looks weirder with the two icons on either side of the year.

image

File: packages/datetime/src/datePickerNavbar.tsx (hunk 4 above)
Comment (giladgray) [reply packages/datetime/src/datePickerNavbar.tsx]: yeah! love a good new feature.

File: packages/datetime/src/datePickerNavbar.tsx (hunk 5 above)
Comment (giladgray) [reply packages/datetime/src/datePickerNavbar.tsx]: DayPicker itself. it's the actual `.DayPicker-whatever` class.

File: packages/datetime/test/dateInputTests.tsx (hunk 6 above)
Comment (giladgray) [reply packages/datetime/test/dateInputTests.tsx]: yeah probs

**Comment by qcz** [comment]
The month picker looks bad in Firefox on Windows:
firefox_2018-08-14_21-47-04
In Edge the icon is still closer to the month name than in Chrome, but at least it does not overlap the name.

File: packages/datetime/src/_datepicker.scss [hunk 7]
```diff
@@ -17,13 +17,24 @@ $header-margin: ($header-height - $pt-input-height) / 2 !default;
   line-height: 1;
 }
 
+.#{$ns}-divider {
```
Comment (giladgray) [inline packages/datetime/src/_datepicker.scss]: TODO: gonna pull this out to its own component in a separate PR, in progress now.

File: packages/datetime/src/datePickerCaption.tsx [hunk 8]
```diff
@@ -44,73 +47,52 @@ export class DatePickerCaption extends React.PureComponent<IDatePickerCaptionPro
...
-            </div>
+            <HTMLSelect
+                iconProps={{ style: { right: this.state.monthRightOffset } }}
+                className={Classes.DATEPICKER_MONTH_SELECT}
+                key="month"
+                minimal={true}
+                onChange={this.handleMonthSelectChange}
+                value={displayMonth}
+                options={monthOptionElements}
+            />
```
Comment (themadcreator) [inline packages/datetime/src/datePickerCaption.tsx]: well this is great

File: packages/datetime/src/datePickerCaption.tsx [hunk 9]
```diff
@@ -123,25 +105,25 @@ export class DatePickerCaption extends React.PureComponent<IDatePickerCaptionPro
...
     private positionArrows() {
         // measure width of text as rendered inside our container element.
-        const monthWidth = Utils.measureTextWidth(
+        const monthTextWidth = measureTextWidth(
             this.displayedMonthText,
             Classes.DATEPICKER_CAPTION_MEASURE,
             this.containerElement,
         );
-        this.setState({ monthWidth });
+        const monthSelectWidth = this.containerElement.firstElementChild.clientWidth;
```
Comment (themadcreator) [inline packages/datetime/src/datePickerCaption.tsx]: reaching into first child element for client width? makes me nervous that we're not doing more null checks

Comment (themadcreator) [reply packages/datetime/src/datePickerCaption.tsx]: also this will trigger a layout

**Pull request merged**